### ISSUE TITLE
Segfault using cookie.parse()

### ISSUE THREAD

**Issue by huguesalary** [lkarsten/libvmod-cookie#28]
Hi, 

I've been trying to use libvmod-cookie (`master` branch, `cc42ac8050`), with varnish 4.0.2, on 2 different systems and I keep bumping into the same issue: varnish crashes. 

I encounter this issue on mac os x, but also on my debian system ( `Linux proxy_varnish4 3.13-1-amd64 #1 SMP Debian 3.13.10-1 (2014-04-15) x86_64 GNU/Linux` with `varnishd (varnish-4.0.2 revision bfe7cd1)`.

In my syslog I find a lot of these:
```
Nov  2 01:15:50 docker kernel: [2782648.550898] traps: varnishd[11449] general protection ip:7f387b9fd77c sp:7f3874dc0110 error:0 in libvmod_cookie.so[7f387b9fc000+3000]
Nov  2 01:16:00 docker kernel: [2782658.858912] varnishd[15102]: segfault at 0 ip 00007f387d7ca784 sp 00007f38748b8110 error 6 in libvmod_cookie.so[7f387d7c9000+3000]
Nov  2 01:22:16 docker kernel: [2783035.163152] varnishd[23528]: segfault at 0 ip 00007f71661fa784 sp 00007f715b9b1110 error 6 in libvmod_cookie.so[7f71661f9000+3000]
Nov  2 01:22:26 docker kernel: [2783044.902367] varnishd[24820]: segfault at 0 ip 00007f71659fa784 sp 00007f715ba08110 error 6 in libvmod_cookie.so[7f71659f9000+3000]
...
```

I suspect the size of my cookies to be the culprit, so I went a ahead and slowly kept increasing the `workspace_session` parameter. Unfortunately this doesn't seem to do anything. My current value is `512000`, which I think is ridiculously high.

Here's an example of my Cookie header triggering the crash:
```
optimizelySegments=%7B%22172841198%22%3A%22direct%22%2C%22172864896%22%3A%22gc%22%2C%22173083061%22%3A%22false%22%7D; optimizelyEndUserId=oeu1414905743840r0.662177964579314; optimizelyBuckets=%7B%7D; __utmt_UA-17580748-5=1; __utma=32229561.1548840885.1414905747.1414905747.1414905747.1; __utmb=32229561.1.10.1414905747; __utmc=32229561; __utmz=32229561.1414905747.1.1.utmcsr=(direct)|utmccn=(direct)|utmcmd=(none)
```

As soon as I make a request with no cookie set (using `curl` for example), it seems to be working.

My VCL looks like this:
```vcl
vcl 4;

import std;
import directors;
import cookie;

sub vcl_recv {
    cookie.parse(req.http.Cookie); #this is the one and only cookie call used in the entire file
    # ...    
}
```

Is there any parameter I should change in my configuration? Or is that an actual bug?

Thanks!

**Comment by lkarsten** [comment]
I've had a deep dive into this, cleaning up and adding a lot of asserts/sanity checks.

ttk's patch was of course right, that debug line should never try to strlen(NULL). Hastily debugging output that made everything worse.

I'm not able to provoke this failure with any of the test cases.

Please remember that storing state across VCL functions in Varnish 4.0 is hacky at best. Run all cookie modifications in vcl_recv and this should not be a factor for you.

Since this is not an actionable bug for me, I'm going to close it. If anyone finds any other crashes, please open a new issue.

**Comment by lkarsten** [comment]
@lgelo re your pull request/gist; I believe I found the reason for it and fixed it in ef0f5dde. The large number of headers made the session workspace overflow, and missing bound checking made one thread trample on some other thread's memory.
